GoRogue/MapGeneration/Region.cs

```diff
@@ -0,0 +1,367 @@
+                ((List<Region>) region.SubRegions).Add(subRegion);
+            }
+            return region;
+        }
+        /// <summary>
+        /// Cuts Points out of a parent region if that point exists in a subregion,
+        /// and cuts Points out of subregions if that point is already in a subregion
+        /// </summary>
+        /// <remarks>
+        /// This favors regions that were added later, like a stack.
+        /// </remarks>
+        public void DistinguishSubRegions()
+        {
+            List<Region> regionsDistinguished = new List<Region>();
+            List<Region> regionsReversed = SubRegions.ToList();
+            regionsReversed.Reverse();
+            foreach (Region region in regionsReversed)
+            {
+                List<Point> removeThese = new List<Point>();
+                foreach (Point point in region.Points.Distinct())
+                {
+                    foreach (Region distinguishedRegion in regionsDistinguished)
+                    {
+                        if (distinguishedRegion.Contains(point))
+                        {
+                            if (region.Contains(point))
+                                removeThese.Add(point);
+                        }
+                    }
+                }
+
+                foreach (var point in removeThese)
+                {
+                    region.Remove(point);
+                }
+                regionsDistinguished.Add(region);
+            }
+        }
+
+        /// <summary>
+        /// Removes a point from the region.
+        /// </summary>
+        /// <param name="point">The point to remove.</param>
+        /// <remarks>
+        /// This removes a point from the inner, outer points, and connections. Does not remove a point
+        /// from a boundary, since those are generated from the corners.
+        /// </remarks>
+        public void Remove(in Point point)
+        {
+            while (_outerPoints.Contains(point))
+                _outerPoints.Remove(point);
+            while (_innerPoints.Contains(point))
+                _innerPoints.Remove(point);
+            while (_connections.Contains(point))
+                _connections.Remove(point);
+        }
+
+        /// <summary>
+        /// Does this region have a subregion with a certain name?
+        /// </summary>
+        /// <param name="name">name of the region to analyze</param>
+        /// <returns>whether the specified region exists</returns>
+        public bool HasRegion(string name) => SubRegions.Where(r => r.Name == name).Count() > 0;
+
+        /// <summary>
+        /// Removes a common point from the OuterPoints of both regions
+        /// and adds it to the Connections of both Regions
+        /// </summary>
+        /// <param name="a">the first region to evaluate</param>
+        /// <param name="b">the second region to evaluate</param>
+        public static void AddConnectionBetween(Region a, Region b, IGenerator rng = null)
```

Comment:
`IGenerator?` due to C# nullable reference types.  This should be giving a compiler warning